src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
+        /// <param name="connection">The connection to return.</param>
+        public void ReturnConnection(HttpConnection connection)
+        {
+            List<CachedConnection> list = _idleConnections;
+            lock (SyncObj)
+            {
+                Debug.Assert(list.Count <= _maxConnections, $"Expected {list.Count} <= {_maxConnections}");
+
+                // Mark the pool as still being active.
+                _usedSinceLastCleanup = true;
+
+                // If there's someone waiting for a connection, simply
+                // transfer this one to them rather than pooling it.
+                if (_waiters != null && _waiters.TryDequeue(out ConnectionWaiter waiter))
+                {
+                    waiter.SetResult(connection);
+                    return;
+                }
+
+                // If the pool has been disposed of, dispose the connection
+                // being returned, as the pool is being deactivated.
+                if (_disposed)
                 {
                     connection.Dispose();
+                    return;
                 }
+
+                // Pool the connection by adding it to the list.
+                list.Add(new CachedConnection(connection));
             }
         }
 
-        private sealed class ConcurrentConnectionSet
+        /// <summary>Disposes the </summary>
+        public void Dispose()
         {
-            private static readonly int s_numSlots = Math.Max(4, RoundUpToPowerOf2(Environment.ProcessorCount));
-            private static readonly int s_slotsMask = s_numSlots - 1;
-            private readonly HashSet<HttpConnection>[] _sets;
+            List<CachedConnection> list = _idleConnections;
+            lock (SyncObj)
+            {
+                if (!_disposed)
+                {
+                    _disposed = true;
+                    list.ForEach(c => c._connection.Dispose());
+                    list.Clear();
+                }
+                Debug.Assert(list.Count == 0, $"Expected {nameof(list)}.{nameof(list.Count)} == 0");
+            }
+        }
 
-            public ConcurrentConnectionSet()
+        /// <summary>
+        /// Removes any unusable connections from the pool, and if the pool
+        /// is then empty and stale, disposes of it.
+        /// </summary>
+        /// <returns>
+        /// true if the pool disposes of itself; otherwise, false.
+        /// </returns>
+        public bool CleanCacheAndDisposeIfUnused()
+        {
+            List<CachedConnection> list = _idleConnections;
+            lock (SyncObj)
             {
-                _sets = new HashSet<HttpConnection>[s_numSlots];
-                for (int i = 0; i < _sets.Length; i++)
+                // Get the current time.  This is compared against each connection's last returned
+                // time to determine whether a connection is too old and should be closed.
+                DateTimeOffset now = DateTimeOffset.Now;
```

Comment:
Nit: I assume that retrieving DateTimeOffset.Now is super cheap, but still, we could do it outside of the lock.